### PR TITLE
Quick site select

### LINKED ISSUE #661: Enable switching sites without logging out
Right now users are required to logout and login again in order to switch sites. We should update the app settings so that tapping the "Primary store" enables choosing a different site.

Android equivalent feature: https://github.com/woocommerce/woocommerce-android/issues/742

### PR DESCRIPTION
**Pull request by mindgraffiti**
Closes #661. 

This PR sets up the quick site select feature, by navigating to Settings and tapping on the switch store row.

http://cld.wthms.co/arJC4W

## To test
1. While already logged in, go to My store > Settings. Check that the section title says "Selected Store". 
2. Tap the Switch Store row. Check that the currently selected store has a checkmark beside it.
3. Choose a different store. Load the other tabs, Orders List and Notifications, to verify the correct data is loading.
4. Log out by visiting My store > Settings Log Out Account.
5. Log in again. Note that the site picker defaults to selecting the first site.
(Try anything else to break it!)

@bummytime @astralbodies ready for review!

Update release notes:

- [x] If there are user facing changes, I have added an item to `RELEASE-NOTES.txt`.

### CODE REVIEW

File: WooCommerce/Classes/ViewRelated/Orders/OrdersViewController.swift [hunk 1]
```diff
@@ -260,6 +274,15 @@ extension OrdersViewController {
...
     }
+
+    /// Default Site Updated Handler
+    ///
+    @objc func defaultSiteWasUpdated() {
+        // dump old orders cache
+        syncingCoordinator.resynchronize()
+        // reload the results controller
+        reloadResultsController()
+    }
```
Comment (mindgraffiti) [inline WooCommerce/Classes/ViewRelated/Orders/OrdersViewController.swift]: I listen for the notification `StoresManagerDidUpdateDefaultSite` and that triggers this method. But my comments (dump old orders cache) do not match what is happening.

File: WooCommerce/Classes/Authentication/Epilogue/StorePickerViewController.swift [hunk 2]
```diff
@@ -343,6 +343,14 @@ extension StorePickerViewController {
...
             dismiss(animated: true) {
                 AppDelegate.shared.authenticatorWasDismissed()
+
+                // grab the tab bar
+                guard let tabBar = AppDelegate.shared.tabBarController else {
+                    return
+                }
+
+                // select My Store
+                tabBar.navigateTo(.myStore, animated: true)
```
Comment (bummytime) [inline WooCommerce/Classes/Authentication/Epilogue/StorePickerViewController.swift]: @mindgraffiti just a quick comment — you can replace everything you added here with `MainTabBarController.switchToMyStoreTab()`

File: WooCommerce/Classes/ViewRelated/Orders/OrdersViewController.swift [hunk 3]
```diff
@@ -85,6 +85,10 @@ class OrdersViewController: UIViewController {
         }
     }
 
+    /// Suppress error warnings when refreshing data in the background
+    ///
+    private var suppressErrorWarning: Bool = false
```
Comment (mindgraffiti) [inline WooCommerce/Classes/ViewRelated/Orders/OrdersViewController.swift]: This seems brittle. Is there a better way to handle the OrdersListViewController's data and state, when it's in the background?

File: WooCommerce/Classes/ViewRelated/Dashboard/Settings/SettingsViewController.swift [hunk 4]
```diff
@@ -216,6 +217,15 @@ private extension SettingsViewController {
         present(alertController, animated: true)
     }
 
+    func changeStoreWasPressed() {
+        StoresManager.shared.resetAuthentication()
+
```
Comment (bummytime) [inline WooCommerce/Classes/ViewRelated/Dashboard/Settings/SettingsViewController.swift]: @mindgraffiti I still have too test this locally, but this is the point where I am thinking we should:
1. Clear all the orders via `OrderAction`resetStoredOrders()`
2. Clear all of the stats data via a **new** action in `StatsAction.resetStoredStats()`

After this we can then show the site picker proceed normally.

File: WooCommerce/Classes/ViewRelated/Orders/OrdersViewController.swift (hunk 3 above)
Comment (bummytime) [reply WooCommerce/Classes/ViewRelated/Orders/OrdersViewController.swift]: @mindgraffiti assuming we clear all of the order data like I suggested above ☝️ , we potentially don't need to do any of the updates here in `OrdersViewController`

**Comment by bummytime** [comment]
Loving what you have so far here @mindgraffiti ! I'm still reviewing this @mindgraffiti, but I wanted to point out a few things to get your thoughts:

## UX flow "trap"

Here is UX flow issue we should probably address. Steps to repro:

1. Open the app
2. Put the phone in airplane mode (to simulate a slow/unresponsive network)
3. Tap settings/switch site
4. See the site picker (you will eventually see a network error notice)
5. User is "stuck" on this screen

napkin 115 02-11-19 9 53 25 am

I know we can kill the app and reopen it to fix things, but I'm thinking we need to consider a way to handle this situation. Ideas:

* The error notice has a "go back" action (in this flow only)? 
* When the site picker is opened via this flow, present it with a standard navbar which has a back button (and still have the existing modal presentation when presented as part of the login flow)
* ???

## Single Site

When the settings screen loads, we should predetermine how many WC sites a user has and not display the "switch site" option if they only have a single site.

File: Yosemite/Yosemite/Stores/StatsStore.swift [hunk 5]
```diff
@@ -39,7 +41,7 @@ public class StatsStore: Store {
 //
 public extension StatsStore  {
 
-    /// Converts a Date into the appropriatly formatted string based on the `OrderStatGranularity`
+    /// Converts a Date into the appropriately formatted string based on the `OrderStatGranularity`
```
Comment (bummytime) [inline Yosemite/Yosemite/Stores/StatsStore.swift]: Always fixing my words! 😆

File: WooCommerce/Classes/ViewRelated/Dashboard/Settings/SettingsViewController.swift [hunk 6]
```diff
@@ -216,6 +225,31 @@ private extension SettingsViewController {
...
+        group.enter()
+        let orderAction = OrderAction.resetStoredOrders {
+            group.leave()
+        }
+        StoresManager.shared.dispatch(orderAction)
+
+        group.notify(queue: .main) {
+            let pickerVC = StorePickerViewController()
+            let loginNavController = LoginNavigationController(rootViewController: pickerVC)
+            self.present(loginNavController, animated: true, completion: nil)
```
Comment (bummytime) [inline WooCommerce/Classes/ViewRelated/Dashboard/Settings/SettingsViewController.swift]: I would add `[weak self] in` to the `group.notify` block and call this as `self?.present(...`

File: WooCommerce/Classes/ViewRelated/Dashboard/Settings/SettingsViewController.swift [hunk 7]
```diff
@@ -131,12 +134,18 @@ private extension SettingsViewController {
...
+    func configureSelectedStore(cell: HeadlineLabelTableViewCell) {
         cell.headline = siteUrl
         cell.body = accountName
         cell.selectionStyle = .none
     }
 
+    func configureSwitchStore(cell: BasicTableViewCell) {
+        cell.accessoryType = .disclosureIndicator
+        cell.selectionStyle = .default
+        cell.textLabel?.text = NSLocalizedString("Switch Store", comment: "This action allows the user to change stores without logging out and logging back in again.")
```
Comment (bummytime) [inline WooCommerce/Classes/ViewRelated/Dashboard/Settings/SettingsViewController.swift]: 👍

File: WooCommerce/Classes/Yosemite/StoresManager.swift [hunk 8]
```diff
@@ -103,6 +103,15 @@ class StoresManager {
         return self
     }
 
+    /// Prepares for changing the selected store and remains Authenticated.
+    ///
+    func resetAuthentication() {
```
Comment (bummytime) [inline WooCommerce/Classes/Yosemite/StoresManager.swift]: Just thinking out loud...I am wondering if this func name could be more precise? Maybe something like `removeDefaultStore`? Any thoughts here?

**Comment by mindgraffiti** [comment]
@bummytime ready for another  👀 

🔢 added unit tests
✅ fixed retain cycle inside of block
➕ renamed method to `removeDefaultStore` (definitely better than reset)
🙅‍♂️ do not display "Switch Store" row unless store.count > 1

As for the UX trap, that's going to take a lot of time to fix. The default site API call is tightly coupled to the Site Picker tableview UI. As in, a new default store is set in storage, an API call is fired to get the new store's data, and the old default store ID in the session manager is erased. So regardless of which type of "go back" method is chosen, they both have the same problem.

I can send you my hacked-together patch to illustrate the separate layers (but it's pretty ugly 😱). 

Thoughts?

**Comment by mindgraffiti** [comment]
Thank you @bummytime and @astralbodies!

**Pull request merged**